### PR TITLE
[x86/Linux] Revise asmhelper.S using macro

### PR DESCRIPTION
**Pull request by parjong**
This commit revises asmhelper.S using macros that inserts CFI
directives.

### CODE REVIEW

**Comment by parjong** [comment]
\CC @seanshpark

File: src/pal/inc/unixasmmacrosx86.inc [hunk 1]
```diff
@@ -35,3 +35,31 @@ C_FUNC(\Name\()_End):
         .global C_FUNC(\Name\()_End)
         LEAF_END \Name, \Section
 .endm
+
+.macro PROLOG_BEG
+        push    ebp
+        mov     ebp, esp
+        .cfi_def_cfa_offset     0
```
Comment (janvorli) [inline src/pal/inc/unixasmmacrosx86.inc]: This is not correct. The CFA by definition is the ESP of the caller right before it made the call. You are setting the CFA here to the location where the EBP was pushed. 
You need to set the offset to 8 (the pushed EBP plus the return address).

File: src/pal/inc/unixasmmacrosx86.inc [hunk 2]
```diff
@@ -35,3 +35,31 @@ C_FUNC(\Name\()_End):
...
+.endm
+
+.macro PROLOG_PUSH Reg
+        push \Reg
+        .cfi_adjust_cfa_offset  -4
+        .cfi_rel_offset \Reg,   0
+.endm
+
+.macro PROLOG_END
+        .cfi_def_cfa_offset     4
```
Comment (janvorli) [inline src/pal/inc/unixasmmacrosx86.inc]: This should be 8, see my comment in the PROLOG_BEG

File: src/pal/inc/unixasmmacrosx86.inc (hunk 1 above)
Comment (janvorli) [inline src/pal/inc/unixasmmacrosx86.inc]: And you need to also define the offset of EBP here
```asm
.cfi_rel_offset ebp, 0
```

File: src/vm/i386/asmhelpers.S [hunk 3]
```diff
@@ -392,8 +399,14 @@ LEAF_END ArrayOpStubTypeMismatchException, _TEXT
 // ------------------------------------------------------------------------------
 //  void STDCALL CallDescrWorkerInternal(CallDescrWorkerParams *  pParams)
 NESTED_ENTRY CallDescrWorkerInternal, _TEXT, NoHandler
+    PROLOG_BEG
+    PROLOG_PUSH eax
```
Comment (janvorli) [inline src/vm/i386/asmhelpers.S]: You don't need to push EAX, EDX and ECX here. They are not expected to be preserved. I have taken a look at a C function compiled by both clang and gcc and they don't preserve these registers either (and use them)

File: src/pal/inc/unixasmmacrosx86.inc (hunk 2 above)
Comment (parjong) [inline src/pal/inc/unixasmmacrosx86.inc]: Thank you for comment. I'll revise those macros as you commented.

File: src/pal/inc/unixasmmacrosx86.inc [hunk 4]
```diff
@@ -35,3 +35,40 @@ C_FUNC(\Name\()_End):
...
+        .cfi_rel_offset \Reg, 0
+.endm
+
+.macro pop_nonvol_reg Reg
+        pop \Reg
+        .cfi_restore \Reg
+.endm
+
+.macro PROLOG_BEG
+        .cfi_def_cfa_offset 4
```
Comment (janvorli) [inline src/pal/inc/unixasmmacrosx86.inc]: I would prefer using less .cfi annotations (the more of them there are, the more work the unwinder has). So I would not use push_nonvol_reg here, but rather make the macro as follows:
```asm
.macro PROLOG_BEG
        push ebp
        .cfi_def_cfa_offset 8
        .cfi_offset ebp, -8
        mov ebp, esp
.endm        
```
That's what clang compiler generates

**Comment by parjong** [comment]
@dotnet-bot test Ubuntu x64 Formatting  please

**Pull request merged**